### ISSUE TITLE
[FR] clang-tidy check for nullptr check when calling delete or free

### ISSUE THREAD

**Issue by FalcoGer** [llvm/llvm-project#118527]
Plenty of people do not know that calling `free`, `std::free` or `delete` on a `nullptr` is a no-op.
If a check for `nullptr` is applied to only such a call, it should be flagged and a fix proposed to remove it.

Possible name could be "readability-redundant-nullptr-check" or it could be in "misc-".

```cpp
int* ptr{nullptr}
if (!ptr)  // check not required
{
    delete ptr;
}
```

All standard checks should be supported `!ptr`, `ptr != nullptr`, `ptr != NULL`, etc
Both extra scopes and single line if checks should be supported.

```cpp
int* ptr{nullptr}
if (!ptr)  // check not required
    delete ptr; // no scope

if (!ptr)  // check not required
{{{{
    delete ptr; // lots of scopes
}}}}
```

Assignment of nullptr after the delete should also trigger this.
Fix should of course retain the assignment.

```cpp
int* ptr{nullptr}
if (!ptr)  // check not required
{
    delete ptr;
    ptr = nullptr; // extra assignment of nullptr. should also check for 0, NULL.
}
```

**Comment by seanm** [comment]
cppcheck has such a warning (if you want to find such issues in your code today).
